[LLVM][DWARF] Refactor code for generating DWARF V5 .debug_names

cmtice opened the pull request:
[LLVM][DWARF]  Refactor code for generating DWARF v5 .debug_names

Refactor the code that uniques the entries and computes the bucket count for the DWARF V5 .debug_names accelerator table.

cmtice committed:
[LLVM][DWARF] Refactor code for generating DWARF V4 .debug_names accelerator table.

Update computeDebugNamesUniqueHashes function to start with lowercase letter,
to use a MutableArrayRef for the hashes parameter instead of SmallVector,
and made uniqueHashCount a reference parameter, to allow the accelerator
table class member to be properly updated. Also removed redundant
namespace qualifier.

cmtice commented:
Could not fix git push/pull/rebase issues with https://github.com/llvm/llvm-project/pull/82264 so had to create this new pull request.  I've tried to address all the review comments from the previous PR:

Update computeDebugNamesUniqueHashes function to start with lowercase letter, to use a MutableArrayRef for the hashes parameter instead of SmallVector, and made uniqueHashCount a reference parameter, to allow the accelerator table class member to be properly updated. Also removed redundant namespace qualifier.

cmtice committed:
[LLVM][DWARF] Refactor code for generating DWARF V5.debug_names

Fix clang format issues.

dwblaikie commented on llvm/include/llvm/BinaryFormat/Dwarf.h:
```diff
@@ -613,6 +613,24 @@ enum AcceleratorTable {
+  uint32_t BucketCount = 0;
+
+  sort(hashes);
+  uniqueHashCount = llvm::unique(hashes) - hashes.begin();
```
Without returning the iterator from `llvm::unique` or otherwise communicating info about `hashes` out to the caller - I guess the caller isn't expected to use their copy of the data structure again? Might be worth detailing in the doc comment that this function effectively consumes the `hashes` array? Or is it somehow usable afterwards?

cmtice replied:
Yes, I want the changes to hashes to make it back to the caller.

I thought making it a MutableArray meant that the changes would get back to the user?  Originally I had it as a SmallVector &, but felipepivezan asked me to change it to a MutableArray instead.  LLVM would not allow me to declare the MutableArray as a reference?

felipepiovezan replied:
> LLVM would not allow me to declare the MutableArray as a reference

MutableArrays are meant to be passed by copy, they are just a "non const ptr" + "size" abstraction.

> I thought making it a MutableArray meant that the changes would get back to the user? 

The changes make it back to the user, but this is not the important distinction between MutableArrayRef and SmallVector. The SmallVector API exposes resizing capabilities, which we are not using inside this function. So you communicate this intent by using the more restricted API.

That said, David's point is that the `erase` functions only moves elements around, it doesn't actually erase them (contrary to what the name suggests). Because of that, the caller of this function can't possibly have a useful vector after this function returns.

cmtice replied:
Ok, I went back and double check the callers; they do NOT expect a usable vector after this call, so that is fine.  However I'm getting a compile time error using the MutableArrayRef, which I could use to help figuring out how to fix: 

third_party/llvm/llvm-project/llvm/lib/CodeGen/AsmPrinter/AccelTable.cpp:43:50: error: non-const lvalue reference to type 'MutableArrayRef<uint32_t>' (aka 'MutableArrayRef<unsigned int>') cannot bind to a value of unrelated type 'SmallVector<uint32_t, 0>' (aka 'SmallVector<unsigned int, 0>')
   43 |       llvm::dwarf::computeDebugNamesUniqueHashes(Uniques, UniqueHashCount);
      |                                                  ^~~~~~~
third_party/llvm/llvm-project/llvm/include/llvm/BinaryFormat/Dwarf.h:618:74: note: passing argument to parameter 'hashes' here
  618 | inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> &hashes,
      |                                                                          ^
1 error generated.

('Uniques', at the call site, line 43, is a SmallVector...)

felipepiovezan replied:
```
618 | inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> &hashes,
```

Did you make it a reference? It should not be a reference

MaskRay replied:
Should the output parameter `uniqueHashCount` be changed to the return value?

cmtice replied:
No, this function calculates and returns two different values:  the bucket count (for which the function is named) and the unique hash count.

felipepiovezan replied:
In hindsight, we should have just returned a pair

cmtice replied:
I'll try to create another PR (probably in a day or two) to fix this. Do you know of any examples I could look at to see how this is done well?

felipepiovezan replied:
> With C++17 structured bindings, returning multiple values via pairs/tuples should really be the norm, rather than using reference parameters as outputs, in my opinion.

FWIW I 100% agree with you, but every now and then this runs into the LLVM's community general hesitancy to use `auto` more aggressively.

> I'll try to create another PR (probably in a day or two) to fix this. Do you know of any examples I could look at to see how this is done well?

Can't think of any example this early, but you could probably do something like:

```
inline std::pair<uint32_t, uint32_t> getDebugNamesBucketAndHashCount((MutableArrayRef<uint32_t> hashes) {
   ...
   return {BucketCount, HashCount};
}
```

And then on the call site you'd normally do something like:

```
auto [BuckerCount, HashCount] = getDebugNamesBucketAndHashCount(...)
```

However, since those are actually member variables, you can't really use a declaration like that. You'll need to store the pair in a temporary variable and assign to the member variables using `.first` and `.second`



```


```

cmtice commented on llvm/include/llvm/BinaryFormat/Dwarf.h:
```diff
@@ -613,6 +613,24 @@ enum AcceleratorTable {
 
+// Uniquify the string hashes and calculate the bucket count for the
+// DWARF v5 Accelerator Table.
+inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> hashes,
```
Well it does return the unique number of hashes, in the second parameter, but I'm Ok with renaming the function...

dwblaikie replied:
Ah. Yeah, naming it after what it returns seems the most legible.

If the intent is to use the uniqued hashes array after the call (not relevant to the current caller, but possibly relevant to the new usage you want to add) maybe the simplest thing is to have the function take a `SmallVectorImpl<uint32_t>&`, and after calling `uique`, use that result to shorten the vector down to the size of the uniqued range (by calling the vector's `resize` function) - then the caller can inspect the vector's size after calling, and doesn't need the secondary out parameter?

cmtice replied:
I've renamed the function as suggested, and added a note in the function comment that it consumes the array parameter.

cmtice committed:
[LLVM][DWARF] Refactor code for generating DWARF v5 .debug_names

Change function name from 'computeDebugNamesUniqueHashes' to
'getDebugNamesBucketCount', and add coment that this function
consumes its input parameter.

cmtice commented:
I think I've addressed all the review comments now? Please let me know if there's anything else I should do for this pull request.

cmtice committed:
[LLVM][DWARF] Refactor code for generating DWARF v5 .debug_names

Clang format fixes.
